core/types/rlp_backwards_compat.libevm_test.go

```diff
@@ -106,3 +110,211 @@ func testHeaderRLPBackwardsCompatibility(t *testing.T) {
+		withdrawals = append(withdrawals, &Withdrawal{
+			Index: rng.Uint64(),
+		})
+	}
+
+	tests := []struct {
+		name        string
+		withdrawals []*Withdrawal
+	}{
+		{
+			name: "with_nil_Withdrawals",
+		},
+		{
+			name: "with_empty_Withdrawals",
+			// As the struct field has an `rlp:"optional"` tag, we need to
+			// differentiate between nil (nothing added to RLP) and an empty
+			// slice (empty list added to RLP).
+			withdrawals: []*Withdrawal{},
+		},
+		{
+			name:        "with_Withdrawals",
+			withdrawals: withdrawals,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			body.Withdrawals = tt.withdrawals
+
+			// The original [Body] doesn't implement [rlp.Encoder] nor
+			// [rlp.Decoder] so we can use a methodless equivalent as the gold
+			// standard.
+			type withoutMethods Body
+			wantRLP, err := rlp.EncodeToBytes((*withoutMethods)(body))
+			require.NoErrorf(t, err, "rlp.EncodeToBytes([%T with methods stripped])", body)
+
+			t.Run("Encode", func(t *testing.T) {
+				got, err := rlp.EncodeToBytes(body)
+				require.NoErrorf(t, err, "rlp.EncodeToBytes(%T)", body)
+				assert.Equalf(t, wantRLP, got, "rlp.EncodeToBytes(%T)", body)
+			})
+
+			t.Run("Decode", func(t *testing.T) {
+				got := new(Body)
+				err := rlp.DecodeBytes(wantRLP, got)
+				require.NoErrorf(t, err, "rlp.DecodeBytes(..., %T)", got)
+
+				opts := cmp.Options{
+					cmpeth.CompareHeadersByHash(),
+					cmpeth.CompareTransactionsByBinary(t),
+				}
+				if diff := cmp.Diff(body, got, opts); diff != "" {
+					t.Errorf("rlp.DecodeBytes(rlp.EncodeToBytes(%T)) diff (-want +got):\n%s", body, diff)
+				}
+			})
+		})
+	}
+}
+
+// cChainBodyExtras carries the same additional fields as the ava-labs/coreth
+// [Body] and implements [BodyHooks] to achieve equivalent RLP {en,de}coding.
+type cChainBodyExtras struct {
+	Version uint32
+	ExtData *[]byte
+}
+
+var _ BodyHooks = (*cChainBodyExtras)(nil)
+
+func (e *cChainBodyExtras) AppendRLPFields(b rlp.EncoderBuffer, _ bool) error {
+	b.WriteUint64(uint64(e.Version))
+	if e.ExtData != nil {
+		b.WriteString(string(*e.ExtData))
+	} else {
+		b.WriteString("")
+	}
+	return nil
+}
+
+func (e *cChainBodyExtras) DecodeExtraRLPFields(s *rlp.Stream) error {
+	if err := s.Decode(&e.Version); err != nil {
+		return err
+	}
+
+	buf, err := s.Bytes()
+	if err != nil {
+		return err
+	}
+	if len(buf) > 0 {
+		e.ExtData = &buf
+	} else {
+		// Respect the `rlp:"nil"` field tag.
+		e.ExtData = nil
+	}
+
+	return nil
+}
+
+func TestBodyRLPCChainCompat(t *testing.T) {
+	// The inputs to this test were used to generate the expected RLP with
+	// ava-labs/coreth. This serves as both an example of how to use [BodyHooks]
+	// and a test of compatibility.
+
+	t.Cleanup(func() {
+		TestOnlyRegisterBodyHooks(NOOPBodyHooks{})
+	})
+
+	to := common.HexToAddress(`decafc0ffeebad`)
+	body := &Body{
+		Transactions: []*Transaction{
+			NewTx(&LegacyTx{
+				Nonce: 42,
+				To:    &to,
```

Comment:
nit could use a little helper function

```go
func ptrTo[T any](x T) *T { return &x }
```

and with

```suggestion
				To:   ptrTo(common.HexToAddress(`decafc0ffeebad`))
```

It's more local/clearer

Reply:
Nice idea. I've added `common.PointerTo()`; thanks.